src/System.Linq.Expressions/src/System/Runtime/CompilerServices/ReadOnlyCollectionBuilder.cs

```diff
@@ -544,24 +583,27 @@ object IEnumerator.Current
 
             public bool MoveNext()
             {
-                if (_version == _builder._version)
+                ReadOnlyCollectionBuilder<T> localBuilder = _builder;
```

Comment:
What problem are these changes fixing?  Some of the cleanup in this PR are good, but please revert the changes that are about throwing different / "better" exceptions.  In general I don't think we should be taking such changes unless there's some other reason for doing so that adds a lot of value, and for those cases, we should have individual issues to discuss the potential changes / breaks.